### DISCUSSION TITLE
Replacing clipTo with clipPath

### DISCUSSION THREAD

**Discussion by dennisideaonce** [fabricjs/fabric.js#7959] [Q&A] [answered]
I am trying to mask an image into a shape using the fabric js clipPath method, and trying to replicate the same behavior of clipPath in two fabric versions those are 1.5.0 and 4.6.0.

Earlier, I was using fabric js version 1.5.0. it was fine (The image fits properly + I can use the controls without any problem), here is the fiddle link:
https://jsfiddle.net/dennisideaonce/czgywasx/1/

Note: The above fiddle uses `canvas.clipTo` method. I need to specifically for the image, not the whole canvas.

Now, I upgraded to fabric js 4.6.0. the masking is not the same as v1.5.0 (The image doesn't fit properly). here is the fiddle link:
https://jsfiddle.net/dennisideaonce/ct3197bo/15/

I am trying to make both the result identical, the purpose is to go to the newer versions of fabric and I want the masking to be done the same way as it was in the older version. I will highly appreciate any help.

**Comment by ShaMan123** [comment, marked as answer]
This is not a bug.
Shocked that you still use v1.5
You should expect much to break leaping 3 majors.

https://jsfiddle.net/tjeadbqm/3/

You are after `shape.globalCompositeOperation='destination-in'`
https://developer.mozilla.org/en-US/docs/Web/API/CanvasRenderingContext2D/globalCompositeOperation

**Reply by dennisideaonce** [reply]
That makes sense, we are using 4.6.0 on production. I tried to achieve the same behavior in the older versions.
https://jsfiddle.net/tjeadbqm/3/ >  you have used `canvas.clipPath = shape`. Is it possible to make it for `imgInstance` not canvas.

Second, I tried the same using `loadSVGFromString` it seems to be working there for me:

https://jsfiddle.net/dennisideaonce/3cgLvbo6/9/

**Comment by dennisideaonce** [comment]
That makes sense, we are using 4.6.0 on production. I tried to achieve the same behavior in the older versions.
https://jsfiddle.net/tjeadbqm/3/ >  you have used `canvas.clipPath = shape`. Is it possible to make it for `imgInstance` not canvas.

Second, I tried the same using `loadSVGFromString` it seems to be working there for me:

https://jsfiddle.net/dennisideaonce/3cgLvbo6/9/

**Reply by ShaMan123** [reply]
Read the docs about clipPath.
`absolutePositioned`
https://jsfiddle.net/qe5fn3Lc/

**Reply by dennisideaonce** [reply]
@ShaMan123 absolutePositioned that's what I was looking for. Thank for the help!

Is it somehow possible if I start dragging the outer frame it should also move the clipped object inside together? I don't want to group them because the inner clipedPath object won't have the controls(as it is coming in V6). Is there any other way apart from a group?

**Comment by dennisideaonce** [comment]
@ShaMan123 `absolutePositioned` that's what I was looking for. Thank for the help!

Is it somehow possible if I start dragging the outer frame it should also move the clipped object inside together? I don't want to group them because the inner clipedPath object won't have the controls(as it is coming in V6). Is there any other way apart from a group?

**Reply by ShaMan123** [reply]
It's hard to understand what you mean...
But this should help
http://fabricjs.com/clipping
Try moving the princess mononoke objects

**Comment by dennisideaonce** [comment]
@ShaMan123 Thanks for the help.

I have achieved moving together, rotation together with the below code (By together I mean mask and image inside the mask):

rotations:

  ```
objectRotating(e) {
    const mainBox = this.canvas.findTarget(e);
    const childBox = mainBox.maskedTo;
    if(!childBox) {
      return true;
    }
    const childCenterPoint = childBox.getCenterPoint();
    const rotationPoint = new fabric.Point(childCenterPoint.x, childCenterPoint.y);
    let originRotationPoint = new fabric.Point(
      mainBox.get("left"),
      mainBox.get("top")
    );
    if (!this.previousAngle) {
      this.previousAngle = mainBox.get("angle");
    }
    let angle = fabric.util.degreesToRadians(
      mainBox.get("angle") - this.previousAngle
    );
    this.previousAngle = mainBox.get("angle");

    let newCoords = fabric.util.rotatePoint(
      rotationPoint,
      originRotationPoint,
      angle
    );

    childBox
      .set({ left: newCoords.x, top: newCoords.y, angle: mainBox.get("angle") })
      .setCoords();
  }
```

Moving mainobject and clipped object:

  ```
moveObject(e) {
    const target = this.canvas.findTarget(e);
    if (target && target.maskedTo) {
      let activeSvg = this.canvas.getActiveObject();
      target.maskedTo.setPositionByOrigin(activeSvg.getCenterPoint(), 'center', 'center');
      target.maskedTo.set('dirty', true);
    }
  }
```


Now, I am facing problems with scaling together the clippedpath image and main object.

**Reply by DineshCodeFlow** [reply]
I am also following this discussion. The problem I am facing is scaling both the objects that are the image and the mask shape together.

**Reply by ShaMan123** [reply]
It's nothing buggy. You need to figure it out.
You need to always take in to account the coordinate plane your object belongs to.
Clip path belongs to the plane created by the object being clipped.
I am guessing you could group the objects (the object that is the clipPath's ghost and the clipped object) and scale with a custom origin point to the center of the object or something like that.
Look into `fabric.util.sendPointToPlane`, `fabric.util.sendObjectToPlane`.

**Reply by ShaMan123** [reply]
A word about belonging to a plane.
An object that belongs to a plane is unaware of the transformations applied to it. It views the plane as not transformed. Meaning if you scale the object the clip path doesn't view itself as scaled.
It is you viewing it as scaled.
